plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,324 @@
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.ldap.LdapName;
+import javax.naming.ldap.Rdn;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCacheWithWeakInvalidateAll;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.containsSpecialCharacters;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.replaceUser;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private final LdapClient client;
+    private final List<String> userBindSearchPatterns;
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> groupBaseDistinguishedName;
+
+    private final Optional<String> groupNameAttribute;
+    private final Optional<String> groupMembershipAttribute;
+    private final Optional<String> groupUserMembershipAttribute;
+    private final String bindDistinguishedName;
+    private final String bindPassword;
+    private final boolean allowUserNotExist;
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.groupNameAttribute = Optional.ofNullable(ldapConfig.getGroupNameAttribute());
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupMembershipAttribute());
+        this.groupUserMembershipAttribute = Optional.ofNullable(ldapConfig.getGroupUserMembershipAttribute());
+        if (this.groupMembershipAttribute.isEmpty() && this.groupUserMembershipAttribute.isEmpty()) {
+            throw new LdapGroupMembershipException("both ldap.group-membership-attribute and "
+                    + "ldap.group-user-membership-attribute cannot be empty");
+        }
+        if (this.groupMembershipAttribute.isPresent() && this.groupNameAttribute.isEmpty()) {
```

Comment:
Can we move this verification to the config class ?

Reply:
We could - but its analogous to the FileProvider - where validation handled in the constructor for the Provider instance itself. The Config classes have no such constructors.

Reply:
We don't need to have a constructor for validation. We could annotate a method with `AssertTrue` or `AssertFalse` to apply corresponding validation. 

For example 
```
    @AssertFalse(message = "Connecting to the LDAP server without SSL enabled requires `ldap.allow-insecure=true`")
    public boolean validateGroupMembership()
    {
        return this.groupMembershipAttribute.isPresent() && this.groupUserMembershipAttribute.isPresent()
    }

Something like this